landscape/client/manager/aptsources.py

```diff
@@ -128,7 +128,10 @@ def _handle_sources(self, ignored, sources):
         for source in sources:
             filename = os.path.join(self.SOURCES_LIST_D,
                                     "landscape-%s.list" % source["name"])
-            with open(filename, "wb") as sources_file:
+            # Servers send unicode, but an upgrade from python2 can get bytes
+            # from stored messages, so we need to handle both.
+            is_unicode = isinstance(source["content"], type(u""))
+            with open(filename, ("w" if is_unicode else "wb")) as sources_file:
```

Comment:
I *want* to rail against this and say "if it's a text file, we should be writing it as a text file", i.e. `with io.open(filename, 'w', encoding=ENCODING) as f: f.write(source['content'].decode(ENCODING))` ... the problem is ENCODING is essentially undefined (deb822 defines it as UTF-8 but that doesn't apply to the older .list style, and besides we don't explicitly know the encoding of `source['content']` either). So, this is probably the best we can do for now.

Reply:
The problem isn't so much the encoding (files are probably ascii as they contain only urls written by landscape server). The issue comes from str getting interpreted as unicode for py3 and bytes for py2, which affected how some of our libraries. I wish we would have taken a slightly different strategy when porting to python3 (which would have highlighted more of those inconsistencies earlier). Unfortunately, this kind of patch is the one with the least impact on what is already released.